source/common/event/dispatcher_impl.cc

```diff
@@ -306,5 +325,26 @@ void DispatcherImpl::touchWatchdog() {
   }
 }
 
+void DispatcherImpl::pushTrackedObject(const ScopeTrackedObject* object) {
+  ASSERT(isThreadSafe());
+  ASSERT(object != nullptr);
+  tracked_object_stack_.push_back(object);
+}
+
+void DispatcherImpl::popTrackedObject(const ScopeTrackedObject* expected_object) {
+  ASSERT(isThreadSafe());
+  ASSERT(expected_object != nullptr);
+  if (tracked_object_stack_.empty()) {
```

Comment:
I think this should probably be `RELEASE_ASSERT(!tracked_object_stack.empty());` It is undefined behavior to `pop_back()` if the container is empty, so I think we should halt even in release builds.